protocol/consistencychecks.go

```diff
@@ -0,0 +1,309 @@
+	case RegistrationType, KeyLookupType:
+		str = msg.DirectoryResponse.(*DirectoryProof).STR
+		// First response
+		if cc.SavedSTR == nil {
+			cc.SavedSTR = str
+			return nil
+		}
+		// FIXME: check whether the STR was issued on time and whatnot.
+		// Maybe it has something to do w/ #81 and client transitioning between epochs.
+		// Try to verify w/ what's been saved
+		if err := cc.verifySTR(str); err == nil {
+			return nil
+		}
+		// Otherwise, expect that we've entered a new epoch
+		if err := cc.verifySTRConsistency(cc.SavedSTR, str); err != nil {
+			return err
+		}
+	}
+
+	// And update the saved STR
+	cc.SavedSTR = str
+	return nil
+}
+
+// verifySTR checks whether the received STR is the same with
+// the SavedSTR using reflect.DeepEqual().
+func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
+	if reflect.DeepEqual(cc.SavedSTR, str) {
+		return nil
+	}
+	return ErrorBadSTR
+}
+
+// verifySTRConsistency checks the consistency between 2 snapshots.
+// It uses the pinned signing key in cc
+// to verify the STR's signature and should not verify
+// the hash chain using the STR stored in cc.
+func (cc *ConsistencyChecks) verifySTRConsistency(savedSTR, str *m.SignedTreeRoot) error {
+	// verify STR's signature
+	if !cc.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+	if str.VerifyHashChain(savedSTR) {
+		return nil
+	}
+
+	// TODO: verify the directory's policies as well. See #115
+	return ErrorBadSTR
+}
+
+func (cc *ConsistencyChecks) checkConsistency(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+	var err error
+	switch requestType {
+	case RegistrationType:
+		err = cc.verifyRegistration(msg, uname, key)
+	case KeyLookupType:
+		err = cc.verifyKeyLookup(msg, uname, key)
+	default:
+		panic("[coniks] Unknown request type.")
+	}
+	return err.(ErrorCode)
+}
+
+func (cc *ConsistencyChecks) verifyRegistration(msg *Response,
+	uname string, key []byte) error {
+	df := msg.DirectoryResponse.(*DirectoryProof)
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	proofType := ap.ProofType()
+	switch {
+	case msg.Error == ErrorNameExisted && proofType == m.ProofOfInclusion && tb == nil:
+	case msg.Error == ErrorNameExisted && proofType == m.ProofOfAbsence:
+	case msg.Error == Success && proofType == m.ProofOfAbsence:
+	default:
+		return ErrorMalformedDirectoryMessage
+	}
+
+	if err := verifyAuthPath(uname, key, ap, str); err != nil {
+		return err
+	}
+
+	return Passed
+}
+
+func (cc *ConsistencyChecks) verifyKeyLookup(msg *Response,
+	uname string, key []byte) error {
+	df := msg.DirectoryResponse.(*DirectoryProof)
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	proofType := ap.ProofType()
+	switch {
+	case msg.Error == ErrorNameNotFound && proofType == m.ProofOfAbsence:
+	// FIXME: This would be changed when we support key changes
+	case msg.Error == Success && proofType == m.ProofOfInclusion && tb == nil:
+	case msg.Error == Success && proofType == m.ProofOfAbsence:
```

Comment:
Same w/ this one `msg.Error == Success && proofType == m.ProofOfAbsence && cc.useTBs`.